### ISSUE TITLE
Support cookie values with comma

### ISSUE THREAD

**Issue by tsu1980** [dotnet/runtime#58773]
The following code will fail.
```
cookieContainer.Add(new Cookie("foo", "ba,r", "/", "example.com"));

---> System.Net.CookieException: The 'Value'='ba,r' part of the cookie is invalid.
```

This means that c# does not allow commas, but major browsers (Chrome, Firefox) do.

Chrome, Edge, and Firefox are the ones that allow commas. The combined market share of these browsers is probably over 90%. This means that **allowing commas is a de facto standard**.

My opinion is to follow de facto standards, not standardization bodies, because our goal is not to follow standards, but to solve real problems.

We need to face this.

----

Similar discussion in Golang: https://github.com/golang/go/issues/7243

**Comment by karelz** [comment]
@tsu1980 I agree we need to be practical with standards. However, I fail to see why HttpClient should be able to send wrong headers and how browser compat plays role here. I would understand the need to accept these headers on the server side.

Are there major servers that are violating the RFC as browsers by sending or requiring commas in cookies?
If not, why should we allow it in .NET?

**Comment by tsu1980** [comment]
I have confirmed that nginx works as follows.

- Can send Set-Cookie header. e.g. `Set-Cookie: foo=ba,r`
- Can parse Cookie header. e.g. `Cookie: foo=ba,r`

**Comment by karelz** [comment]
@tsu1980 does it send cookie headers with comma by default? Or using some standard configuration?
Are you aware of any real-world sites which use comma separated cookies?

**Comment by tsu1980** [comment]
@karelz 
It's not my server, so I don't know about the server settings. However, I don't think there is such a setting in nginx.

> Are you aware of any real-world sites which use comma separated cookies?

Yes. That's why I reported this issue.

I also think that such web applications are badly behaved and rare. However, such web applications do exist in real, because it works without any problems🙁

**Comment by karelz** [comment]
@tsu1980 can you share which server endpoint it is? Is there a way to learn from owners of the server which version it is and if it is misconfigured?
Does the server require comma-separation and refuses the (correct) semi-colon separation?

**Comment by tsu1980** [comment]
@karelz I'm sorry, but I can't share the site.

Instead, I tried this with Apache and PHP at my hand.

PHP's setcookie function escapes the value, so I added a cookie by the Apache configuration file.

#### 1) Setup apache config
Apache config:
```
#cookie-with-comma.localhost
<VirtualHost *:80>
    DocumentRoot "<path-to>/cookie-with-comma/public_html"
    ServerName cookie-with-comma.localhost

    <Directory "<path-to>/cookie-with-comma/public_html">
        DirectoryIndex index.php
        AllowOverride All
        Require all granted
    </Directory>

    <Location /action.php>
        Header append Set-Cookie "ValueWithInvalidChars=abc, xyz; path=/"
        Header append Set-Cookie "ValueWithInvalidChars2=abc,; xyz; path=/"
    </Location>
</VirtualHost>
```

#### 2) Setup PHP files
index.php:
```
<p>It's working.</p>
```

action.php:
```
<p>Set cookies.</p>
```

#### 3) Access to http://cookie-with-comma.localhost/action.php in Chrome
Server response of /action.php:
```
HTTP/1.1 200 OK
Date: Wed, 08 Sep 2021 11:13:06 GMT
Server: Apache/2.4.20 (Win32) OpenSSL/1.0.2g PHP/7.2.10
X-Powered-By: PHP/7.2.10
Set-Cookie: ValueWithInvalidChars=abc, xyz; path=/, ValueWithInvalidChars2=abc,; xyz; path=/
Content-Length: 27
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8
X-DNS-Prefetch-Control: off
```

Cookies on http://cookie-with-comma.localhost/:
![cookie-with-commas2](https://user-images.githubusercontent.com/724468/132498003-b01e7d3a-428d-482d-8854-ac7d7294f652.png)

**Comment by karelz** [comment]
@tsu1980 in your test case, the server returns comma-separated cookies, but there is no need for client to construct and send comma-separated cookies. Is that correct?
Does HttpClient have problems with parsing comma-separated cookies in response from the server?

**Comment by ManickaP** [comment]
I haven't read the whole discussion here, but this might be relevant: https://github.com/dotnet/runtime/issues/42856 since we do have a bug in concatenating cookie values...

**Comment by tsu1980** [comment]
Yep, HttpClient([CookieContainer.SetCookies](https://docs.microsoft.com/en-us/dotnet/api/system.net.cookiecontainer.setcookies?view=net-5.0)) cannot parse the following cookies in the same way as Chrome.

```
Set-Cookie: ValueWithInvalidChars=abc, xyz; path=/, ValueWithInvalidChars2=abc,; xyz; path=/
```

**Comment by karelz** [comment]
@tsu1980 can you please confirm if `HttpClient` has problems to parse response with comma-separated cookies?
I don't think we need cookie APIs to be able to parse such input from user input. Though it would be nice to have a way for reverse-proxy scenarios (cc @alnikola @MihaZupan).

**Comment by tsu1980** [comment]
Yeah i can confirm that `HttpClient` cannot parse response with comma-separated cookies. Because I've encountered it.

It will look like this.

#### Server response cookies:
```
Set-Cookie: ValueWithInvalidChars=abc, xyz; path=/
```

#### c# CookieContainer.SetCookies result:
- ValueWithInvalidChars=abc
- xyz=

#### Chrome parse result:
- ValueWithInvalidChars=abc, xyz

----
I think this(parsing Set-Cookie/Cookie header) problem is reported as #42856 already.

**Comment by ManickaP** [comment]
@karelz we might reconsider milestone (Future -> 7.0) for #42856. It seems like proper cookie handling could solve the problem reported here and IMHO it shouldn't be hard to implement. Moreover, we would be becoming compliant with RFC instead of breaking it as we do now.

**Comment by scalablecory** [comment]
I agree this and #42856 make sense for 7.0, if we see enough upvotes.

**Comment by istefanou** [comment]
Has this been scheduled for any shortly upcoming version?

**Comment by ManickaP** [comment]
Hi, unfortunately no, we haven't fixed it yet. But I've marked both cookie issues (this and #42856) as "up for grabs" and we'd gladly accept a community PR for them.

**Comment by devsko** [comment]
The following code works fine in all current .NET Core versions (3.1, 5, 6, 7) as well as in .NET 4.8
```csharp
   var domain = new Uri("http://example.com");
   var container = new CookieContainer();
   container.Add(domain, new Cookie("name", "\"abc,def\""));
```
A `HttpClient` renders this as `name="abc,def"`.

@istefanou 
> But I gave it a try and it doesn't so I assume that the functionality is supposed to be there right?

Can you give me some more details about what is not working?

**Comment by istefanou** [comment]
Hi @devsko I am using .NET 4.7.2
It seems that this indeed works with the proper escaping. (facepalm)

However my actual use-case is the following. Can this be achieved?

Cookie myCookie = new Cookie();
myCookie.Name = "MyCookie";
myCookie.Value = "{\"Shape\":1,\"Color\":1}";
cookieContainer.Add(uri, myCookie);

The value of the cookie is supposed to be {"Shape":1,"Color":1}

**Comment by devsko** [comment]
I am not sure what you are trying to accomplish but this issue is mainly about `HttpClient` and receiving cookies in .NET 7. It looks like you maybe want to use ASP.NET Core.

**Comment by gioping** [comment]
Only work if (e.g ABC="aaa,bbb,ccc")
It still invalid when cookie value without "  (e.g  ABC=aaa,bbb,ccc )

**Comment by devsko** [comment]
Hi @ManickaP.  Basically, I am done with this change. It contains the following:

- Support RFC 6265 by default

  - Adding `CookieVariant.Rfc6265` (I do not consider this a breaking change) and set it as the default for all cookies.
  - Enforce the rules for cookie name and value. This can be considered a breaking change because it allows less than before.
  - Parse cookie headers compliant to RFC6265 by default (no `COMMA` as separator, no obsoleted tokens like `Port`, `Comment` etc., no multiple cookies in response header)
  - Remove the logic that allows all characters when a value is `DQUOTE`d because there is no such specification.
    
- Relax the rules to accept what major browsers support

  - Allow `COMMA` and `SP` in name and value.
  - Allow cookies with empty name. Yes, this is supported by all browsers i tested.

- Be backward compatible

  - Restart header parsing in "legacy mode" When a `[$]Version` token is found (AFAIK this should never happen except during unit tests).
  - Change `Cookie.Variant` when non RFC6265 properties are set.
  - Try not to change anything about domain / path matching.

- Kind of rewrite of `CookieParser` for obvious reasons.

Unfortunately it took me a while to realize that correctly and reliably detecting the need to fallback to legacy mode adds a significant overhead to response header parsing. Given that this case should be - at least - VERY rare this is not acceptable. (BTW `HttpListener` produces non RFC6265 compliant response headers)

To go forward i see 2 possible solutions
  1. Remove back compat completely at least from header parsing as suggested by #40328.
  2. Make RFC6265 header parsing Opt-out. This would however require an API Proposal.

Please let me know how to proceed. Thanks.

**Comment by istefanou** [comment]
Hello @devsko and @ManickaP, hope you are doing well.

Is there any update on this?
As mentioned in my comments below I am seeing functionality issues with cookies on websites using
Cookies which have values like {"Shape":1,"Color":6}

Thanks in advance

**Comment by ManickaP** [comment]
@istefanou  could you try with .NET 7 preview version? The fix @devsko did was merged this year, so it's only in .NET 7.

Unfortunately, I haven't had a chance to properly check [@devsko's comment](https://github.com/dotnet/runtime/issues/58773#issuecomment-1096463881) and https://github.com/dotnet/runtime/issues/40328, so I'm not 100% sure this should be closed as resolved since we already have the "RFC 6265 for Cookies" issue.

**Comment by heinrich-ulbricht** [comment]
Still happening in .NET 8. Have issues setting string cookies values that are serialized JSON.

**Comment by ManickaP** [comment]
Technically, having comma in cookie value is not allowed: https://httpwg.org/specs/rfc6265.html#rfc.section.4.1.1
You should percent encode it.

**Comment by ManickaP** [comment]
>  Screenshot from Developer Tools Cookie view for a Confluence Cloud site.

Just a thought, but the server might be percent-encoding and the client percent-decoding and showing the decoded value in UI. Have you try to see what actually goes over the wire, e.g. by wireshark?

**Comment by heinrich-ulbricht** [comment]
@ManickaP Good idea. I had a look at the raw response in Fiddler and visiting atlassian.com also seems to set the cookie. Looks JSON to me:

![Image](https://github.com/user-attachments/assets/e9d410ff-7971-4be4-b244-112d0f24daeb)

**Comment by ManickaP** [comment]
So some suggestions that might work for you as a workaround:
- try percent encode the value when adding the cookie and see if the server can accept it and decode it
- you can add the header manually (not like a cookie) via `TryAddWithoutValidation` on the request headers
